src/pip/_internal/wheel.py

```diff
@@ -773,52 +773,84 @@ def _contains_egg_info(
-    ephemeral cache.
-
-    :param cache_available: whether a cache directory is available for the
-        should_unpack=True case.
-
-    :return: True or False to build the requirement with ephem_cache=True
-        or False, respectively; or None not to build the requirement.
+    Return whether a wheel should be built based on the requirement
+    characteristics, and if a wheel is ultimately needed or not.
     """
     if req.constraint:
         # never build requirements that are merely constraints
-        return None
+        return False
+
     if req.is_wheel:
-        if not should_unpack:
+        if need_wheel:
             logger.info(
                 'Skipping %s, due to already being wheel.', req.name,
             )
-        return None
-    if not should_unpack:
-        # i.e. pip wheel, not pip install;
-        # return False, knowing that the caller will never cache
-        # in this case anyway, so this return merely means "build it".
-        # TODO improve this behavior
+        # never build a requirement that is already a wheel
         return False
 
-    if req.editable or not req.source_dir:
-        return None
+    if need_wheel:
+        return True
+
+    if req.editable:
+        # don't build an editable if no wheel is needed
+        return False
+
+    if not req.source_dir:
+        # TODO explain what no source_dir means
+        return False
 
     if "binary" not in format_control.get_allowed_formats(
             canonicalize_name(req.name)):
         logger.info(
             "Skipping bdist_wheel for %s, due to binaries "
             "being disabled for it.", req.name,
         )
-        return None
+        return False
+
+    return True
+
+
+def should_use_ephemeral_cache(
+    req,  # type: InstallRequirement
+    format_control,  # type: FormatControl
+    cache_available,  # type: bool
+):
+    # type: (...) -> Optional[bool]
+    """
+    Return whether to build an InstallRequirement object using the
+    ephemeral cache.
+
+    :param cache_available: whether a cache directory is available for the
+        should_unpack=True case.
+
+    :return: True or False to build the requirement with ephem_cache=True
+        or False, respectively; or None not to build the requirement.
+    """
+    if req.editable:
+        # don't cache editable requirements because they can
+        # be locally modified
+        return True
+
+    if not req.source_dir:
+        # TODO explain what no source_dir means
+        return True
+
+    if "binary" not in format_control.get_allowed_formats(
+            canonicalize_name(req.name)):
+        # TODO explain this
+        return True
 
     if req.link and req.link.is_vcs:
-        # VCS checkout. Build wheel just for this run.
+        # VCS checkout. Build wheel just for this run
+        # unless it points to an immutable commit hash in which
+        # case it can be cached.
+        # TODO if req.link.is_vcs_immutable: return False
```

Comment:
I would simplify this comment because e.g. the method may not be called `is_vcs_immutable()`. I think something like the following would be enough to indicate what is planned (and be more accurate): "TODO: cache against the commit hash if we can."

Reply:
Done